File: Simplify Path.java
// Leetcode Problem 71: Simplify Path
// JAVA CODE
class Solution {
    public String simplifyPath(String path) {
        Deque<String> stack = new ArrayDeque<>();

        for (String part : path.split("/")) {
            if (part.equals("") || part.equals(".")) {
                continue;
            } else if (part.equals("..")) {
                if (!stack.isEmpty()) {
                    stack.pollLast();
                }
            } else {
                stack.offerLast(part);
            }
        }

        if (stack.isEmpty()) {
            return "/";
        }

        StringBuilder result = new StringBuilder();

        while (!stack.isEmpty()) {
            result.append("/").append(stack.pollFirst());
        }

        return result.toString();
    }
}

File: Simplify Path.cpp
// Leetcode Problem 71: Simplify Path
// C++ CODE
class Solution {
public:
    string simplifyPath(string path) {
        vector<string> stack;
        string part;
        stringstream ss(path);

        while (getline(ss, part, '/')) {
            if (part.empty() || part == ".") {
                continue;
            } else if (part == "..") {
                if (!stack.empty()) {
                    stack.pop_back();
                }
            } else {
                stack.push_back(part);
            }
        }

        if (stack.empty()) {
            return "/";
        }

        string result;
        for (string &dir : stack) {
            result += "/" + dir;
        }

        return result;
    }
};